src/app/chat/chat-view/chat-view.component.ts

```diff
@@ -1,66 +1,110 @@
   selector: 'chat-view',
   templateUrl: './chat-view.component.html',
   styleUrls: ['./chat-view.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class ChatViewComponent implements OnInit, OnDestroy {
 
   @ViewChild('chatContent') chatContent: any;
-  public channel;
+  public channel: ChannelByNameQuery.ChannelByName;
   private routeParamsSub;
+  private messagesSub;
   public model = { message: undefined };
-  public messages = [
-    {
-      content: 'kentak is pitushky',
-      creationTime: new Date().getTime() - (Math.random() * 100000000),
-      avatar: 'http://dreamicus.com/data/face/face-01.jpg',
-      user: { name: 'GushBasar' }
-    },
-    {
-      content: 'kentak is kentak',
-      creationTime: new Date().getTime() - (Math.random() * 100000000),
-      avatar: 'http://dreamicus.com/data/face/face-01.jpg',
-      user: { name: 'GushBasar' }
-    },
-    {
-      content: 'fried chicken is kentak',
-      creationTime: new Date().getTime() - (Math.random() * 100000000),
-      avatar: 'http://dreamicus.com/data/face/face-01.jpg',
-      user: { name: 'GushBasar' }
-    }
-  ];
+  private pageMessagesCount = 100;
+  private readonly pagePercentLoadMoreTrigger = 0.45;
+  private isFirstLoad = true;
+  public messages;
+  private pagePixelLength: number;
 
-  constructor(private route: ActivatedRoute) {
+  constructor(private route: ActivatedRoute, public chatService: ChatService, private cd: ChangeDetectorRef) {
   }
 
   ngOnInit() {
-    // TODO handle direct too
-    this.routeParamsSub = this.route.params.subscribe(params => this.channel = params['id']);
+    this.routeParamsSub = this.route.params.subscribe(params => {
+      if (this.messagesSub) {
+        this.messagesSub.unsubscribe();
+      }
+
+      const url: any = this.route.url;
+      const isDirect = url.value[0].path === 'direct';
+      const channelName = params['id'];
+
+      this.chatService.getChannelByName(channelName, isDirect).subscribe((channelData) => {
+        this.channel = channelData.data.channelByName;
+        if (this.channel === null) {
+          return;
+        }
+
+        Observable.fromEvent(this.chatContent.nativeElement, 'scroll').subscribe(() => {
+          if (this.chatContent.nativeElement.scrollTop < this.pagePixelLength * this.pagePercentLoadMoreTrigger) {
+            if (!this.chatService.isLoadingMoreMessages()) {
+              this.chatService.loadMoreMessages(this.channel.id, this.pageMessagesCount);
+            }
+          }
+        });
+
+        const messagesObs = this.chatService.getMessages(this.channel.id, this.pageMessagesCount);
+
+        this.messagesSub = messagesObs.subscribe(({ data }) => {
+          const oldScrollHeight = this.chatContent.nativeElement.scrollHeight;
+          this.messages = data.messages.messagesArray;
+          if (this.isFirstLoad || (!this.chatService.isLoadingMoreMessages() && this.isScrolledToBottom())) {
+            setTimeout(() => {
```

Comment:
why setTimeout? arent a zone triggered from the message arrival?
and if so you can use setImmediate